azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -182,12 +182,26 @@ def run(self, goal_state, ext_gs_updated):
             if not self._updater.is_retrieved_version_allowed_to_update(agent_family):
                 return
             self._updater.log_new_agent_update_message()
-            self._updater.purge_extra_agents_from_disk()
             agent = self._updater.download_and_get_new_agent(self._protocol, agent_family, goal_state)
-            if agent.is_blacklisted or not agent.is_downloaded:
-                msg = "Downloaded agent version is in bad state : {0} , skipping agent update".format(
-                    str(agent.version))
+
+            # We are not updating the agent if it is not downloaded properly or agent is blacklisted and already attempted 3 times.
+            # For some reason, if agent is blackslisted in previous attempts(<3), we clear the error state to make agent as good agent.
+            # If we allow to update, we increment the update attempt count.
+            if not agent.is_downloaded:
+                msg = "Agent: {0} is not downloaded properly, skipping agent update".format(str(agent.version))
                 raise AgentUpdateError(msg)
+            elif agent.get_update_attempt_count() >= 3 and agent.is_blacklisted:
```

Comment:
how does this interact with RSM updates? won't this prevent updates when RSM tries the update again?

Reply:
That was the intention, right otherwise we will be in loop. If transient failure, this may help. If it's permanent, what's the point of honoring RSM request.

After first RSM request, every new GS is an update. So, we don't know which GS driven by RSM.

Reply:
It would kill further rsm retries too.

Reply:
Have you talked this over with the RSM team?

Reply:
I haven't but what is that we can do otherwise to fix this issue?

I will check with them anyway.